start.py

```diff
@@ -0,0 +1,121 @@
+# limitations under the License.
+
+import logging
+import re
+from os.path import expanduser
+
+from clusterdock.models import Cluster, Node
+from clusterdock.utils import wait_for_condition
+
+logger = logging.getLogger('clusterdock.{}'.format(__name__))
+
+DEFAULT_NAMESPACE = 'clusterdock'
+DEFAULT_OPERATING_SYSTEM = 'centos6.6'
+
+KDC_ACL_FILENAME = '/var/kerberos/krb5kdc/kadm5.acl'
+KDC_CONF_FILENAME = '/var/kerberos/krb5kdc/kdc.conf'
+KERBEROS_VOLUME_DIR = '/etc/clusterdock/kerberos'
+KDC_KEYTAB_FILENAME = '{}/clusterdock.keytab'.format(KERBEROS_VOLUME_DIR)
+KDC_KRB5_CONF_FILENAME = '/etc/krb5.conf'
+
+
+def main(args):
+    image = '{}/{}/topology_nodebase:{}'.format(args.registry, args.namespace or DEFAULT_NAMESPACE,
+                                                args.operating_system or DEFAULT_OPERATING_SYSTEM)
+    nodes = [Node(hostname=hostname, group='nodes', image=image)
+             for hostname in args.nodes]
+
+    kerberos_volume_dir = args.kerberos_config_directory.replace('~', expanduser('~'))
+    kdc_image = '{}/{}/topology_nodebase_kerberos:{}'.format(args.registry, args.namespace or DEFAULT_NAMESPACE,
+                                                             args.operating_system or DEFAULT_OPERATING_SYSTEM)
+    kdc_hostname = args.kdc_node[0]
+    kdc_node = Node(hostname=kdc_hostname, group='kdc', image=kdc_image,
+                    volumes=[{kerberos_volume_dir: KERBEROS_VOLUME_DIR}])
+    cluster = Cluster(kdc_node, *nodes)
+    cluster.start(args.network)
+
+    logger.info('Updating KDC configurations ...')
+    realm = cluster.network.upper()
+    # Update configurations
+    krb5_conf_data = kdc_node.get_file(KDC_KRB5_CONF_FILENAME)
+    kdc_node.put_file(KDC_KRB5_CONF_FILENAME,
+                      re.sub(r'EXAMPLE.COM', realm,
+                             re.sub(r'example.com', cluster.network,
+                                    re.sub(r'kerberos.example.com', r'{}.{}'.format(kdc_hostname, cluster.network),
+                                           krb5_conf_data))))
+    kdc_conf_data = kdc_node.get_file(KDC_CONF_FILENAME)
+    kdc_node.put_file(KDC_CONF_FILENAME,
+                      re.sub(r'EXAMPLE.COM', realm,
+                             kdc_conf_data.replace(r'[kdcdefaults]',
+                                                   '[kdcdefaults]\n max_renewablelife = 7d\n max_life = 1d')))
+    acl_data = kdc_node.get_file(KDC_ACL_FILENAME)
+    kdc_node.put_file(KDC_ACL_FILENAME, re.sub(r'EXAMPLE.COM', realm, acl_data))
+
+    kdc_commands = [
+        'kdb5_util create -s -r {realm} -P kdcadmin'.format(realm=realm),
+        'kadmin.local -q "addprinc -pw {adminpw} admin/admin@{realm}"'.format(adminpw='acladmin', realm=realm)
+    ]
+
+    logger.info('Starting KDC ...')
+    # Add the following commands before starting kadmin daemon etc.
+    if args.kerberos_principals:
+        principal_list = ['{}@{}'.format(primary, realm) for primary in args.kerberos_principals.split(',')]
```

Comment:
If each element of args.kerberos_principals is actually the primary, should we rename the arg?

Reply:
actually principal is the right naming convention. so changed primary->principal